utils/ujson-cmp.man
.\" Manpage for ujson-verify
.\" Contact dan@ultramarin.se to correct errors or types.
.TH ujson-cmp 1 "" "" "User Commands"


.SH NAME
ujson-cmp \- Compare two JSON documents


.SH SYNOPSIS
.B ujson-cmp
[OPTIONS...] [FILE_1] [FILE_2]


.SH DESCRIPTION
ujson-cmp parses two JSON documents and compares the instances for equality. If the JSON instances are equal, ujson-cmp prints 'equal' to standard output and exits with code 0. If they are not equal, ujson-cmp prints 'not equal' to standard output and exits with code 1. On errors, for example parsing errors, an error message is printed to standard error and the exit code is 1.


.SH OPTIONS
.TP
.B -r, --relaxed
Parse JSON documents in relaxed mode.
.TP
.B -q, --quiet
Silent mode, don't write anything to standard output.
.TP
.B -v, --version
Print version and exit.
.TP
.B -h, --help
Print help and exit.


.SH NOTES
By default JSON documents are parsed using the official JSON syntax. But libujson also support a more relaxed format.
.PP
The following is allowed in JSON documents when relaxed format is used:

.nf
- C-style comments.
- String values can be split up into more than one string separated by whitespace or comments (like in C/C++).
- Object definitions are allowed to end with a separator(,).
- Array items are allowed to end with a separator(,).
- In object definitions, a member name can be an 'identifier' instead of a string enclosed by double quotes.
  An identifier is an UTF-8 encoded name in the following format: [_a-zA-Z][_a-zA-Z0-9]*
  Exceptions are: true, false, and null (case insensitive).
  Those names are reserved and not allowed to be used as identifiers (i.e. without enclosing double quotes).

.PP
Example of a JSON document that is allowed in relaxed mode:

.EX
{
    /*
     * An invalid JSON syntax in strict mode, but allowed in relaxed mode.
     */
    name: "relaxed document",
    value: 42,
    desc: "Some text here",
    relaxed_member_name: "A string",
    "strict_member_name": "Another string",

    // The following object member name is reserved and must be enclosed by double quotes
    "null": 0,

    long_string: "This is a string value on " // This can be used to
                 "more than one line that, "  // make it more readable.
                 "when parsed, will be one "
                 "long string value.",

    // Note that the last array item ends with a separator
    list_of_values: [1, 2, 3, 4, 5, 6, 7,],

    // Also, the last object member can end with a separator
    last_member: true,
}
.EE

.SH SEE ALSO
ujson-print(1) ujson-get(1) ujson-verify(1) ujson-patch(1)


.SH AUTHOR
Dan Arrhenius (dan@ultramarin.se)


utils/ujson-cmp.cpp
/*
 * Copyright (C) 2021,2022 Dan Arrhenius <dan@ultramarin.se>
 *
 * This file is part of ujson.
 *
 * ujson is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include <ujson.hpp>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cstring>
#include <cerrno>
#include <unistd.h>
#include <getopt.h>


using namespace std;

static constexpr const char* prog_name = "ujson-cmp";

struct appargs_t {
    bool relaxed;
    bool quiet;
    string filename[2];

    appargs_t () {
        relaxed = false;
        quiet = false;
    }
};


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static void print_usage_and_exit (std::ostream& out, int exit_code)
{
    out << endl
        << "Check if two JSON instances are equal." << endl
        << endl
        << "Usage: " << prog_name << " [OPTIONS] [FILE_1] [FILE_2]" << endl
        << endl
        << "Options:" <<endl
        << "  -r, --relaxed    Parse JSON documents in relaxed mode." << endl
        << "  -q, --quiet      Silent mode, don't write anything to standard output." << endl
        << "  -v, --version    Print version and exit." << endl
        << "  -h, --help       Print this help message and exit." << endl
        << endl;
        exit (exit_code);
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static void parse_args (int argc, char* argv[], appargs_t& args)
{
    struct option long_options[] = {
        { "relaxed", no_argument, 0, 'r'},
        { "quiet",   no_argument, 0, 'q'},
        { "version", no_argument, 0, 'v'},
        { "help",    no_argument, 0, 'h'},
        { 0, 0, 0, 0}
    };
    const char* arg_format = "rqvh";

    while (1) {
        int c = getopt_long (argc, argv, arg_format, long_options, NULL);
        if (c == -1)
            break;
        switch (c) {
        case 'r':
            args.relaxed = true;
            break;
        case 'q':
            args.quiet = true;
            break;
        case 'v':
            std::cout << prog_name << ' ' << PACKAGE_VERSION << std::endl;
            exit (0);
            break;
        case 'h':
            print_usage_and_exit (std::cout, 0);
            break;
        default:
            exit (1);
            break;
        }
    }
    if (optind < argc)
        args.filename[0] = argv[optind++];
    if (optind < argc)
        args.filename[1] = argv[optind++];

    if (args.filename[0].empty() || args.filename[1].empty()) {
        if (!args.quiet)
            cerr << "Missing filename(s)" << endl;
        exit (1);
    }
    if (optind < argc) {
        if (!args.quiet)
            cerr << "Too many arguments" << endl;
        exit (1);
    }
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
int main (int argc, char* argv[])
{
    appargs_t opt;

    parse_args (argc, argv, opt);

    // Read json files
    //
    ifstream ifs[2];
    string json_desc[2];
    for (auto i=0; i<2; ++i) {
        // Open
        ifs[i].open (opt.filename[i]);
        if (ifs[i].fail()) {
            cerr << "Can't open file \"" << opt.filename[i] << "\" - " << strerror(errno) << endl;
            exit (1);
        }
        // Read
        json_desc[i] = string ((istreambuf_iterator<char>(ifs[i])), istreambuf_iterator<char>());
        if (ifs[i].fail()) {
            cerr << "Error reading file \"" << opt.filename[i] << "\" - " << strerror(errno) << endl;
            exit (1);
        }
        ifs[i].close ();
    }

    // Parse json files
    //
    ujson::Json j;
    ujson::jvalue instance[2];
    for (auto i=0; i<2; ++i) {
        instance[i] = j.parse_string (json_desc[i], !opt.relaxed);
        if (!instance[i].valid()) {
            if (!opt.quiet)
                cerr << "Error parsing " << opt.filename[i] << ": " << j.error() << endl;
            exit (1);
        }
    }

    // Check if the instances are equal
    //
    int retval = 0;
    if (instance[0] == instance[1]) {
        if (!opt.quiet)
            cout << "equal" << endl;
    }else{
        if (!opt.quiet)
            cout << "not equal" << endl;
        retval = 1;
    }

    return retval;
}
